stdlib/src/utils/static_tuple.mojo

```diff
@@ -257,7 +257,11 @@ struct InlineArray[
 
     # Fields
     alias type = __mlir_type[
-        `!pop.array<`, size.value, `, `, Self.ElementType, `>`
+        `!pop.array<`,
+        size.value,
+        `, `,
+        UnsafeMaybeUninitialized[Self.ElementType],
+        `>`,
```

Comment:
I think I understand the intention here, but I think this may not actually be required.

I think the state we want to aim for is:

* `InlineArray.__init__(unsafe_uninitialized)` is replaced with:

    ```mojo
    @staticmethod
    fn unsafe_uninitialized() -> InlineArray[UnsafeMaybeUninitialized[T]]:
        ...
    ```

* All the remaining `InlineArray.__init__(...)` constructors should (and currently do) always fully initialize the `InlineArray`. 
  
   (This is okay to do without making the `pop.array` storage type use `UnsafeMaybeUninitialized` because the compiler knows that inside an `__init__` method `self` and the fields of `self` won't be fully initialized anyway.

With this approach, all the methods off `InlineArray` will be able to assume one of the following two things is true:

1. The current `InlineArray` is fully initialized
2. The current `InlineArray` may not be fully initialized, but its element types are definitely wrapped in `UnsafeMaybeUninitialized[..]`, so that's okay.

We'd then also want to add this constructor for the programmer to "assert" that an uninitialized `InlineArray` is now fully initialized.

```
struct InlineArray[T]:
    fn __init__(
        inout self,
        owned unsafe_assume_initialized: InlineArray[UnsafeMaybeUninitialized[T]]
    ):
        # transmute the argument `InlineArray` into one that is now considered fully initialized
        ...
```

@gabrieldemarmiesse  What do you think of this?

I think it would eliminate the need for some of the changes that had to be made in the current iteration of this PR throughout the `InlineArray` implementation.

Reply:
Using `UnsafeMaybeUninitialized` inside `InlineArray` is indeed not necessary, but it was intentional to modify the struct like this. The idea was to move away from pointers and use only `UnsafeMaybeUninitialized` which has a restricted use compared to pointers (notably, no pointer arithmetic). We still want to carefully control the lifecycle, and `UnsafeMaybeUninitialized` allows it.

But sure I can change it back and use plain pointers again, though I believe the code is less readable.

Reply:
I think a pointer-based implementation is still my personal preference, since it limits the number of types involved. But I see the merits of your argument too, and I don't want to hold up this excellent work on a small difference in perspective like that 🙂

So I'm going to resync this and merge it, and let this shake out—we can always revisit and cleanup later if there's a readability or maintainability problem in practice.

And I agree, we can / should add `@staticmethod fn unsafe_uninitialized()` in a separate PR.

Reply:
I don't mind reverting one day to pointers if we see that this is causing some issues down the road :)

Reply:
I'll revert it to see if it fixes the IR